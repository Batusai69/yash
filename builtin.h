/* Yash: yet another shell */
/* builtin.h: built-in commands */
/* (C) 2007-2012 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef YASH_BUILTIN_H
#define YASH_BUILTIN_H

#include <stddef.h>


typedef int main_T(int argc, void **argv)
    __attribute__((nonnull));

typedef enum builtintype_T {
    BI_SPECIAL, BI_SEMISPECIAL, BI_REGULAR,
} builtintype_T;

typedef struct builtin_T {
    main_T *body;
    builtintype_T type;
#if YASH_ENABLE_HELP
    const char *help_text, *syntax_text;
    const struct xgetopt_T *options;
#endif
} builtin_T;


extern void init_builtin(void);
extern const builtin_T *get_builtin(const char *name)
    __attribute__((pure));

extern int mutually_exclusive_option_error(wchar_t opt1, wchar_t opt2);
extern bool validate_operand_count(size_t count, size_t min, size_t max);
extern int insufficient_operands_error(size_t min_required_operand_count);
extern int too_many_operands_error(size_t max_accepted_operand_count);
extern int special_builtin_syntax_error(int exitstatus);

extern int print_builtin_help(const wchar_t *name)
    __attribute__((nonnull));
extern bool print_shopts(bool include_normal_options);
extern bool print_option_list(const struct xgetopt_T *options)
    __attribute__((nonnull));

extern int true_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern const char colon_help[], colon_syntax[], true_help[], true_syntax[];

extern int false_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern const char false_help[], false_syntax[];

extern int help_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern const char help_help[], help_syntax[];


#endif /* YASH_BUILTIN_H */


/* vim: set ts=8 sts=4 sw=4 noet tw=80: */
