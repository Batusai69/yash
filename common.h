/* Yash: yet another shell */
/* common.h: defines symbols common to all sources. */
/* (C) 2007-2010 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* This file should be included at the very first in every source file. */


#ifndef YASH_COMMON_H
#define YASH_COMMON_H

#if HAVE_CONFIG_H
# include "config.h"
#endif

#ifndef _POSIX_C_SOURCE
# define _POSIX_C_SOURCE 200112L
#endif
#ifndef _XOPEN_SOURCE
# define _XOPEN_SOURCE 600
#endif

#ifndef __GNUC__
# define __attribute__(ignore)
#endif

#ifdef __CYGWIN__
# undef __STRICT_ANSI__
#endif

#if HAVE_GETTEXT
# define gt(MSGID)  ((const char *) gettext(MSGID))
# define Ngt(MSGID) MSGID
# if HAVE_NGETTEXT
#  define ngt(MSGID,MSGS,MSGP,N) ((const char *) ngettext(MSGS, MSGP, N))
# else
#  define ngt(MSGID,MSGS,MSGP,N) gt(MSGID)
# endif
#else
# define gt(MSGID)              MSGID
# define Ngt(MSGID)             MSGID
# define ngt(MSGID,MSGS,MSGP,N) MSGID
#endif

#define ARGV(i) ((wchar_t *) argv[i])

#endif /* YASH_COMMON_H */


/* vim: set ts=8 sts=4 sw=4 noet tw=80: */
