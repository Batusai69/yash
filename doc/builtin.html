<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 9.1.0" />
<meta name="description" content="This page describes common properties of built-ins of yash." />
<title>Built-in commands</title>
<link rel="stylesheet" href="./asciidoc.css" type="text/css" />


</head>
<body class="article">
<div id="header">
<h1>Built-in commands</h1>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph"><p><dfn>Built-in commands</dfn> are commands that are implemented in the shell and are
executed by the shell without starting external programs.</p></div>
<div class="paragraph"><p>See the <a href="index.html#builtins">table of contents</a> for the list of the
built-ins.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="types">Types of built-in commands</h2>
<div class="sectionbody">
<div class="paragraph"><p>Yash has several types of built-in commands described below.</p></div>
<div class="paragraph"><p><dfn>Special built-in commands</dfn> are the most important kinds of built-in
commands.
They are executed regardless of whether the corresponding external commands
exist or not. Results of variable assignments that occur in a
<a href="syntax.html#simple">simple command</a> that invokes a special built-in last
after the command has finished. Moreover, in the
<a href="posix.html">POSIXly-correct mode</a>, a <a href="interact.html">non-interactive</a>
shell immediately exits with a non-zero exit status when an error occurs in a
simple command executing a special built-in.</p></div>
<div class="paragraph"><p><dfn>Mandatory built-in commands</dfn> and <dfn>elective built-in commands</dfn> are
similar to special built-ins in that they do not need external commands to be
executed but differ in that they may be overridden by
<a href="exec.html#function">functions</a>.
While mandatory built-ins are always available, you cannot use elective
built-ins in the <a href="posix.html">POSIXly-correct mode</a> because POSIX only
reserves their names without defining their behavior.</p></div>
<div class="paragraph"><p>An <dfn>extension built-in command</dfn> is a built-in that is not mentioned in
POSIX. Like an elective built-in, it can be executed without an external
command when the <a href="posix.html">POSIXly-correct mode</a> is off. However, when
the POSIXly-correct mode is on, the shell works as if the built-in does not
exist.</p></div>
<div class="paragraph"><p><dfn>Substitutive built-in commands</dfn> work on behalf of external commands
<a href="exec.html#search">found in PATH</a>. These built-ins improve execution speed
by bypassing invocation overheads for external programs.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="argsyntax">Syntax of command arguments</h2>
<div class="sectionbody">
<div class="paragraph"><p>In this section we explain common rules about command arguments.
The built-in commands of yash follow the rules unless otherwise stated.</p></div>
<div class="paragraph"><p>There are two types of command arguments. One is options and the other is
operands. An option is an argument that starts with a hyphen (<code>-</code>) and
changes the way the command behaves. Some options take arguments. An operand
is an argument that is not an option and specifies objects the command
operates on.</p></div>
<div class="paragraph"><p>If you specify more than one option to a command, the order of the options are
normally not significant. The order of operands, however, affects the command
behavior.</p></div>
<div class="paragraph"><p>An option is either a single-character option or a long option.
A single-character option is identified by one alphabetic character. A long
option is identified by multiple alphabetic characters. The POSIX standard only
prescribes single-character options, so in the <a href="posix.html">POSIXly-correct
mode</a> you cannot use long options.</p></div>
<div class="paragraph"><p>A single-character option is composed of a hyphen followed by a letter. For
example, <code>-a</code> is a single-character option. A single-character option that
takes an argument requires the argument to be just after the option name.</p></div>
<div class="exampleblock">
<div class="title">Example 1. The set built-in and single-character options</div>
<div class="content">
<div class="paragraph"><p>For the set built-in, <code>-m</code> is a single-character option that does not take an
argument and <code>-o</code> is one that takes an argument.</p></div>
<div class="ulist"><ul>
<li>
<p>
<code>set -o errexit -m</code>
</p>
</li>
<li>
<p>
<code>set -oerrexit -m</code>
</p>
</li>
</ul></div>
<div class="paragraph"><p>In these two command lines, <code>errexit</code> is the argument to the <code>-o</code> option.</p></div>
</div></div>
<div class="paragraph"><p>In the second example above, the <code>-o</code> option and its argument are combined
into a single command line argument.
The POSIX standard deprecates that style and any POSIX-conforming applications
must specify options and their arguments as separate command line arguments,
although yash accepts both styles.</p></div>
<div class="paragraph"><p>You can combine single-character options that do not take arguments into a
single command line argument. For example, the three options <code>-a</code>, <code>-b</code> and
<code>-c</code> can be combined into <code>-abc</code>.</p></div>
<div class="paragraph"><p>A long option is composed of two hyphens followed by an option name.
For example, <code>--long-option</code> is a long option.
You can omit some last characters of a long option name as long as it is not ambiguous.
For example, you can use <code>--long</code> instead of <code>--long-option</code>
if there is no other options beginning with <code>--long</code>.
Like a single-character option, a long option that takes an argument requires
the argument to be a command line argument just after the option name or to be
specified in the same command line argument as the option name,
separated by an equal sign (<code>=</code>).</p></div>
<div class="exampleblock">
<div class="title">Example 2. The fc built-in and long options</div>
<div class="content">
<div class="paragraph"><p>For the fc built-in, <code>--quiet</code> is a long option that does not take an argument
and <code>--editor</code> is one that takes an argument.</p></div>
<div class="ulist"><ul>
<li>
<p>
<code>fc --editor vi --quiet</code>
</p>
</li>
<li>
<p>
<code>fc --editor=vi --quiet</code>
</p>
</li>
</ul></div>
<div class="paragraph"><p>In these command lines, <code>vi</code> is the argument to the <code>--editor</code> option.</p></div>
</div></div>
<div class="paragraph"><p>Arguments that are not options (nor arguments to them) are interpreted as
operands. The POSIX standard requires all options should be specified before
any operands. Therefore, in the <a href="posix.html">POSIXly-correct mode</a>, any
arguments that come after the first operand are interpreted as operands (even
if they look like options). If not in the POSIXly-correct mode, you can
specify options after operand.</p></div>
<div class="paragraph"><p>Regardless of whether the shell is in the POSIXly-correct mode or not, an
argument that is just composed of two hyphens (<code>--</code>) can be used as a
separator between options and operands. All command line arguments after the
<code>--</code> separator are interpreted as operands, so you can specify operands that
start with a hyphen correctly using the separator.</p></div>
<div class="exampleblock">
<div class="title">Example 3. Options and operands to the set built-in</div>
<div class="content">
<div class="ulist"><ul>
<li>
<p>
<code>set -a -b -- -c -d</code>
</p>
</li>
</ul></div>
<div class="paragraph"><p>In this example, <code>-a</code> and <code>-b</code> are options and <code>-c</code> and <code>-d</code> are operands. The
<code>--</code> separator itself is neither an option nor an operand.</p></div>
</div></div>
<div class="paragraph"><p>Regardless of whether the shell is in the POSIXly-correct mode or not, an
argument that is just composed of a single hyphen (<code>-</code>) is interpreted as an
operand.</p></div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated
 2023-04-22 01:39:14 JST
</div>
</div>
</body>
</html>
