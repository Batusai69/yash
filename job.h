/* Yash: yet another shell */
/* job.h: job control */
/* © 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef JOB_H
#define JOB_H

#include <stdbool.h>
#include <stddef.h>
#include <sys/types.h>


/* ジョブ／プロセスの状態 */
typedef enum jobstatus_T {
    JS_RUNNING, JS_STOPPED, JS_DONE,
} jobstatus_T;

/* ジョブを構成するプロセスの情報 */
typedef struct process_T {
    pid_t             pr_pid;          /* プロセス ID */
    enum jobstatus_T  pr_status;       /* プロセスの状態 */
    int               pr_statuscode;
    wchar_t          *pr_name;         /* 表示用のプロセス名 */
} process_T;
/* pr_pid が 0 のとき、そのプロセスは fork せずに実行できたことを示す。このとき
 * そのコマンドの終了コードが pr_statuscode に入る。
 * pr_pid が正数のとき、それはプロセス ID である。このときの pr_statuscode は
 * waitpid が返した status の値である。 */

/* ジョブの情報 */
typedef struct job_T {
    pid_t             j_pgid;          /* プロセスグループ ID */
    enum jobstatus_T  j_status;        /* ジョブの状態 */
    bool              j_statuschanged; /* 状態変化を報告していないなら true */
    bool              j_loop;          /* ループパイプかどうか */
    size_t            j_pcount;        /* j_procs の要素数 */
    struct process_T  j_procs[];       /* 各プロセスの情報 */
} job_T;
/* ジョブ制御無効時、ジョブのプロセスグループ ID はシェルと同じになるので、
 * j_pgid は 0 となる。 */


/* アクティブジョブのジョブ番号 */
#define ACTIVE_JOBNO 0

/* プロセスがシグナルによって終了した場合は、シグナル番号に TERMSIGOFFSET を
 * 加えた値がそのプロセスの終了ステータスになる。
 * bash/zsh/dash ではこの値は 128、ksh では 256 である。 */
#define TERMSIGOFFSET 384

extern void init_job(void);

extern void set_active_job(job_T *job)
    __attribute__((nonnull));
extern void add_job(void);
extern void remove_job(size_t jobnumber);
extern void remove_all_jobs(void);

extern void do_wait(void);
extern void wait_for_job(size_t jobnumber, bool return_on_stop);

extern int calc_status_of_job(const job_T *job)
    __attribute__((pure,nonnull));


#endif /* JOB_H */


/* vim: set ts=8 sts=4 sw=4 noet: */
