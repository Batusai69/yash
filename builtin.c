/* Yash: yet another shell */
/* builtin.c: built-in commands */
/* (C) 2007-2011 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#include "common.h"
#include <assert.h>
#include <errno.h>
#if HAVE_GETTEXT
# include <libintl.h>
#endif
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#if YASH_ENABLE_ALIAS
# include "alias.h"
#endif
#include "builtin.h"
#include "exec.h"
#include "hashtable.h"
#if YASH_ENABLE_HISTORY
# include "history.h"
#endif
#include "job.h"
#include "option.h"
#include "path.h"
#include "sig.h"
#include "strbuf.h"
#include "util.h"
#include "variable.h"
#include "xfnmatch.h"
#include "yash.h"
#if YASH_ENABLE_PRINTF
# include "builtins/printf.h"
#endif
#if YASH_ENABLE_TEST
# include "builtins/test.h"
#endif
#if YASH_ENABLE_ULIMIT
# include "builtins/ulimit.h"
#endif
#if YASH_ENABLE_LINEEDIT
# include "lineedit/complete.h"
# include "lineedit/keymap.h"
#endif


/* Rules about built-in commands:
 * - `argc' passed to a built-in is at least one; The command name is given in
 *   `argv[0]'.
 * - `argv' may be rearranged and the values of the argument strings may be
 *   changed in the built-in. However, the argument strings may not be `free'd
 *   or `realloc'ed.
 * - Built-ins may sleep or wait, but cannot be stopped. */


/* A hashtable from names of built-ins (const char *) to built-in info
 * structures (const builtin_T *). */
static hashtable_T builtins;

/* Initializes `builtins' */
void init_builtin(void)
{
    assert(builtins.capacity == 0);
    ht_initwithcapacity(&builtins, hashstr, htstrcmp, 53);

#if YASH_ENABLE_HELP
# define DEFBUILTIN(name,func,type,help) \
    do {                                                                      \
	static const builtin_T bi = { func, type, help, };                    \
	ht_set(&builtins, name, &bi);                                         \
    } while (0)
#else
# define DEFBUILTIN(name,func,type,help) \
    do {                                                                      \
	static const builtin_T bi = { func, type, };                          \
	ht_set(&builtins, name, &bi);                                         \
    } while (0)
#endif

    /* defined in "builtin.c" */
    DEFBUILTIN(":", true_builtin, BI_SPECIAL, colon_help);
    DEFBUILTIN("true", true_builtin, BI_SEMISPECIAL, true_help);
    DEFBUILTIN("false", false_builtin, BI_SEMISPECIAL, false_help);
#if YASH_ENABLE_HELP
    DEFBUILTIN("help", help_builtin, BI_REGULAR, help_help);
#endif

    /* defined in "option.c" */
    DEFBUILTIN("set", set_builtin, BI_SPECIAL, set_help);

    /* defined in "path.c" */
    DEFBUILTIN("cd", cd_builtin, BI_SEMISPECIAL, cd_help);
    DEFBUILTIN("pwd", pwd_builtin, BI_SEMISPECIAL, pwd_help);
    DEFBUILTIN("hash", hash_builtin, BI_REGULAR, hash_help);
    DEFBUILTIN("umask", umask_builtin, BI_SEMISPECIAL, umask_help);

    /* defined in "alias.c" */
#if YASH_ENABLE_ALIAS
    DEFBUILTIN("alias", alias_builtin, BI_SEMISPECIAL, alias_help);
    DEFBUILTIN("unalias", unalias_builtin, BI_SEMISPECIAL, unalias_help);
#endif

    /* defined in "variable.c" */
    DEFBUILTIN("typeset", typeset_builtin, BI_REGULAR, typeset_help);
    DEFBUILTIN("export", typeset_builtin, BI_SPECIAL, typeset_help);
    DEFBUILTIN("readonly", typeset_builtin, BI_SPECIAL, typeset_help);
#if YASH_ENABLE_ARRAY
    DEFBUILTIN("array", array_builtin, BI_REGULAR, array_help);
#endif
    DEFBUILTIN("unset", unset_builtin, BI_SPECIAL, unset_help);
    DEFBUILTIN("shift", shift_builtin, BI_SPECIAL, shift_help);
    DEFBUILTIN("getopts", getopts_builtin, BI_SEMISPECIAL, getopts_help);
    DEFBUILTIN("read", read_builtin, BI_SEMISPECIAL, read_help);
#if YASH_ENABLE_DIRSTACK
    DEFBUILTIN("pushd", pushd_builtin, BI_REGULAR, pushd_help);
    DEFBUILTIN("popd", popd_builtin, BI_REGULAR, popd_help);
    DEFBUILTIN("dirs", dirs_builtin, BI_REGULAR, dirs_help);
#endif

    /* defined in "sig.c" */
    DEFBUILTIN("trap", trap_builtin, BI_SPECIAL, trap_help);
    DEFBUILTIN("kill", kill_builtin, BI_SEMISPECIAL, kill_help);

    /* defined in "job.c" */
    DEFBUILTIN("jobs", jobs_builtin, BI_SEMISPECIAL, jobs_help);
    DEFBUILTIN("fg", fg_builtin, BI_SEMISPECIAL, fg_help);
    DEFBUILTIN("bg", fg_builtin, BI_SEMISPECIAL, bg_help);
    DEFBUILTIN("wait", wait_builtin, BI_SEMISPECIAL, wait_help);
    DEFBUILTIN("disown", disown_builtin, BI_REGULAR, disown_help);

    /* defined in "history.c" */
#if YASH_ENABLE_HISTORY
    DEFBUILTIN("fc", fc_builtin, BI_SEMISPECIAL, fc_help);
    DEFBUILTIN("history", history_builtin, BI_REGULAR, history_help);
#endif

    /* defined in "exec.c" */
    DEFBUILTIN("return", return_builtin, BI_SPECIAL, return_help);
    DEFBUILTIN("break", break_builtin, BI_SPECIAL, break_help);
    DEFBUILTIN("continue", break_builtin, BI_SPECIAL, continue_help);
    DEFBUILTIN("eval", eval_builtin, BI_SPECIAL, eval_help);
    DEFBUILTIN(".", dot_builtin, BI_SPECIAL, dot_help);
    DEFBUILTIN("exec", exec_builtin, BI_SPECIAL, exec_help);
    DEFBUILTIN("command", command_builtin, BI_SEMISPECIAL, command_help);
    DEFBUILTIN("type", command_builtin, BI_REGULAR, type_help);
    DEFBUILTIN("times", times_builtin, BI_SPECIAL, times_help);

    /* defined in "yash.c" */
    DEFBUILTIN("exit", exit_builtin, BI_SPECIAL, exit_help);
    DEFBUILTIN("suspend", suspend_builtin, BI_REGULAR, suspend_help);

    /* defined in "builtins/ulimit.c" */
#if YASH_ENABLE_ULIMIT
    DEFBUILTIN("ulimit", ulimit_builtin, BI_REGULAR, ulimit_help);
#endif

    /* defined in "builtins/printf.c" */
#if YASH_ENABLE_PRINTF
    DEFBUILTIN("echo", echo_builtin, BI_REGULAR, echo_help);
    DEFBUILTIN("printf", printf_builtin, BI_REGULAR, printf_help);
#endif

    /* defined in "builtins/test.c" */
#if YASH_ENABLE_TEST
    DEFBUILTIN("test", test_builtin, BI_REGULAR, test_help);
    DEFBUILTIN("[", test_builtin, BI_REGULAR, test_help);
#endif

    /* defined in "lineedit/complete.c" */
#if YASH_ENABLE_LINEEDIT
    DEFBUILTIN("complete", complete_builtin, BI_REGULAR, complete_help);
#endif

    /* defined in "lineedit/keymap.c" */
#if YASH_ENABLE_LINEEDIT
    DEFBUILTIN("bindkey", bindkey_builtin, BI_REGULAR, bindkey_help);
#endif

#undef DEFBUILTIN
}

/* Returns the built-in command of the specified name or NULL if not found. */
const builtin_T *get_builtin(const char *name)
{
    return ht_get(&builtins, name).value;
}

#if YASH_ENABLE_HELP

/* Prints description of the specified built-in to the standard output.
 * Returns Exit_SUCCESS if the built-in was found and the help was printed.
 * Returns Exit_FAILURE if the built-in was not found or an error occurred. */
int print_builtin_help(const wchar_t *name)
{
    char *mbsname = malloc_wcstombs(name);
    const builtin_T *bi = get_builtin(mbsname);
    free(mbsname);
    if (bi == NULL) {
	xerror(0, Ngt("no such built-in `%ls'"), name);
	return Exit_FAILURE;
    }

    for (const char **help = bi->help; *help != NULL; help++)
	if (!xprintf("%s", gt(*help)))
	    return Exit_FAILURE;
    return Exit_SUCCESS;
}

#endif /* YASH_ENABLE_HELP */

#if YASH_ENABLE_LINEEDIT

/* Generates completion candidates for built-in command names matching the
 * pattern. The CGT_SBUILTIN, CGT_SSBUILTIN, and CGT_RBUILTIN flags in
 * `compopt->type' specify what candidate to generate. The other flags are
 * ignored. */
/* The prototype of this function is declared in "lineedit/complete.h". */
void generate_builtin_candidates(const le_compopt_T *compopt)
{
    if (!(compopt->type & CGT_BUILTIN))
	return;

    le_compdebug("adding built-in command name candidates");
    if (!le_compile_cpatterns(compopt))
	return;

    size_t i = 0;
    kvpair_T kv;
    while ((kv = ht_next(&builtins, &i)).key != NULL) {
	le_candgentype_T type;
	switch (((const builtin_T *) kv.value)->type) {
	    case BI_SPECIAL:      type = CGT_SBUILTIN;   break;
	    case BI_SEMISPECIAL:  type = CGT_SSBUILTIN;  break;
	    case BI_REGULAR:      type = CGT_RBUILTIN;   break;
	    default:          assert(false);
	}
	if (!(compopt->type & type))
	    continue;

	if (le_match_comppatterns(compopt, kv.key))
	    le_new_candidate(CT_COMMAND,
		    malloc_mbstowcs(kv.key), NULL, compopt);
    }
}

#endif /* YASH_ENABLE_LINEEDIT */


/* The ":"/"true" built-in. */
int true_builtin(
	int argc __attribute__((unused)), void **argv __attribute__((unused)))
{
    return EXIT_SUCCESS;
}

/* The "false" built-in. */
int false_builtin(
	int argc __attribute__((unused)), void **argv __attribute__((unused)))
{
    return EXIT_FAILURE;
}

#if YASH_ENABLE_HELP

const char *colon_help[] = { Ngt(
"colon - null utility\n"
), Ngt(
"\t: [arg...]\n"
), Ngt(
"The colon built-in does nothing. Any arguments are ignored.\n"
), Ngt(
"Note that arguments are expanded and redirections are performed as usual.\n"
"The colon and true built-ins have the same effect, but colon is a special\n"
"built-in while true is a semi-special.\n"
), NULL };

const char *true_help[] = { Ngt(
"true - return true value\n"
), Ngt(
"\ttrue\n"
), Ngt(
"The true built-in does nothing successfully.\n"
"Any arguments are ignored and the exit status is always zero.\n"
), Ngt(
"Note that arguments are expanded and redirections are performed as usual.\n"
"The colon and true built-ins have the same effect, but colon is a special\n"
"built-in while true is a semi-special.\n"
), NULL };

const char *false_help[] = { Ngt(
"false - return false value\n"
), Ngt(
"\tfalse\n"
), Ngt(
"The false built-in does nothing unsuccessfully.\n"
"Any arguments are ignored and the exit status is always non-zero.\n"
), NULL };


/* The "help" built-in. */
int help_builtin(int argc, void **argv)
{
    const struct xgetopt_T *opt;
    xoptind = 0;
    while ((opt = xgetopt(argv, help_option, 0)) != NULL) {
	switch (opt->shortopt) {
	    case L'-':
		goto print_help;
	    default:
		fprintf(stderr, gt("Usage:  help command...\n"));
		return Exit_ERROR;
	}
    }

    if (xoptind == argc)
print_help:
	return print_builtin_help(ARGV(0));

    while (xoptind < argc)
	print_builtin_help(ARGV(xoptind++));
    return (yash_error_message_count == 0) ? Exit_SUCCESS : Exit_FAILURE;
}

const char *help_help[] = { Ngt(
"help - print usage of built-ins\n"
), Ngt(
"\thelp command...\n"
), Ngt(
"The help built-in prints a description of the specified built-in <command>s.\n"
), NULL };

#endif /* YASH_ENABLE_HELP */


/* vim: set ts=8 sts=4 sw=4 noet tw=80: */
