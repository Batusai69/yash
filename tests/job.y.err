===== error =====
fg: --no-such-option: invalid option
Usage:  fg [job]
fg: %100: no such job
fg: %no_such_job: no such job
fg: too many operands
Usage:  fg [job]
fg: job control disabled
bg: --no-such-option: invalid option
Usage:  bg [job...]
bg: %100: no such job
bg: %no_such_job: no such job
bg: job control disabled
