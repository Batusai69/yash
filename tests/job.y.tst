# job.y.tst: yash-specific test of job control
# vim: set ft=sh ts=8 sts=4 sw=4 noet:


echo ===== fg bg suspend =====

set -m +o curstop

sleep   `echo  0`&
fg
$INVOKE $TESTEE -c 'suspend; echo 1'
$INVOKE $TESTEE -c 'suspend; echo 2'
$INVOKE $TESTEE -c 'suspend; echo 3'
jobs
fg %2 3 '%? echo 1'

$INVOKE $TESTEE -c 'suspend; echo 4'
jobs %%
bg
wait %%
$INVOKE $TESTEE -c 'suspend'
$INVOKE $TESTEE -c 'suspend'
$INVOKE $TESTEE -c 'suspend; echo 7'
bg %1 %2 
wait
bg '${'
fg '? echo 7' >/dev/null

$INVOKE $TESTEE -imc --norc 'echo 1; suspend; echo 2'
kill -l $?
bg >/dev/null
wait %1
kill -l $?
fg >/dev/null

set +m


echo ===== error =====
echo ===== error ===== >&2

$INVOKE $TESTEE -m <<\END
fg --no-such-option
echo fg no-such-option $?
exit 100 &
fg >&- 2>/dev/null
END
echo fg output error $?
$INVOKE $TESTEE -m <<\END
fg %100
echo fg no-such-job 1 $?
fg %no_such_job
echo fg no-such-job 2 $?
set --posix
exit 101 & exit 102 &
fg %1 %2
echo fg too many args $?
END
fg
echo fg +m $?

$INVOKE $TESTEE -m <<\END
bg --no-such-option
echo bg no-such-option $?
while kill -0 $$; do sleep 1; done 2>/dev/null&
bg >&- 2>/dev/null
kill %1
END
echo bg output error $?
set -m
bg %100
echo bg no-such-job 1 $?
bg %no_such_job
echo bg no-such-job 2 $?
set +m
bg
echo bg +m $?

