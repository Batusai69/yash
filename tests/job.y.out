===== fg bg suspend =====
[1] sleep $(echo  0)
[1]   Stopped(SIGSTOP)     ${INVOKE} ${TESTEE} -c 'suspend; echo 1'
[2] - Stopped(SIGSTOP)     ${INVOKE} ${TESTEE} -c 'suspend; echo 2'
[3] + Stopped(SIGSTOP)     ${INVOKE} ${TESTEE} -c 'suspend; echo 3'
[2] ${INVOKE} ${TESTEE} -c 'suspend; echo 2'
2
[3] ${INVOKE} ${TESTEE} -c 'suspend; echo 3'
3
[1] ${INVOKE} ${TESTEE} -c 'suspend; echo 1'
1
[1] + Stopped(SIGSTOP)     ${INVOKE} ${TESTEE} -c 'suspend; echo 4'
[1] ${INVOKE} ${TESTEE} -c 'suspend; echo 4'
4
[1] ${INVOKE} ${TESTEE} -c 'suspend'
[2] ${INVOKE} ${TESTEE} -c 'suspend'
[3] ${INVOKE} ${TESTEE} -c 'suspend; echo 7'
7
1
STOP
TTOU
2
===== test =====
test 1 0
test 2 0
test 3 1
test 4 1
===== error =====
fg no-such-option 2
fg output error 100
fg no-such-job 1 1
fg no-such-job 2 1
fg too many args 2
fg +m 1
bg no-such-option 2
bg output error 0
bg no-such-job 1 1
bg no-such-job 2 1
bg +m 1
===== signals =====
INT ignored
TERM ignored
QUIT ignored
INT
TERM
QUIT
TSTP ignored
TTOU
TSTP
1
2
===== signals +i -m F =====
ABRT
ALRM
BUS
FPE
HUP
ILL
INT
KILL
PIPE
QUIT
SEGV
TERM
USR1
USR2
CHLD 0
URG 0
===== signals -i -m F =====
ABRT
ALRM
BUS
FPE
HUP
ILL
KILL
PIPE
SEGV
USR1
USR2
CHLD 0
URG 0
===== signals +i -m T =====
ABRT
ALRM
BUS
FPE
HUP
ILL
INT
KILL
PIPE
QUIT
SEGV
TERM
USR1
USR2
CHLD 0
URG 0
TSTP
TSTP!
TSTP 0
TTIN
TTIN!
TTIN 0
TTOU
TTOU!
TTOU 0
STOP
STOP!
STOP 0
===== signals -i -m T =====
ABRT
ALRM
BUS
FPE
HUP
ILL
KILL
PIPE
QUIT
SEGV
TERM
USR1
USR2
CHLD 0
URG 0
INT
TSTP
TSTP!
TSTP 0
TTIN
TTIN!
TTIN 0
TTOU
TTOU!
TTOU 0
STOP
STOP!
STOP 0
